File: Ahorn/entities/SpiderBaker.jl
module SpiderBaker

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/SpiderBaker" SpiderBaker(x::Integer, y::Integer, detectionRange::Integer=100, health::Integer=2, startFriendly::Bool=false, webY::Integer=-80)

const placements = Ahorn.PlacementDict(
    "neutral" => Ahorn.EntityPlacement(SpiderBaker),
    "friendly" => Ahorn.EntityPlacement(SpiderBaker)
)

function Ahorn.selection(entity::SpiderBaker)
    return Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::SpiderBaker, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end

File: Loenn/entities/spiderBaker.lua
local utils = require("utils")
local drawableSprite = require("structs.drawable_sprite")

local spiderBaker = {}

spiderBaker.name = "MaggyHelper/SpiderBaker"
spiderBaker.depth = -100
spiderBaker.fieldInformation = {
    health = {
        minimumValue = 1
    },
    detectionRange = {
        minimumValue = 0
    },
    webY = {
        fieldType = "integer"
    }
}
spiderBaker.placements = {
    {
        name = "neutral",
        data = {
            health = 2,
            detectionRange = 100,
            webY = -80,
            startFriendly = false
        }
    },
    {
        name = "friendly",
        data = {
            health = 2,
            detectionRange = 100,
            webY = -80,
            startFriendly = true
        }
    }
}

function spiderBaker.sprite(room, entity)
    local sprite = drawableSprite.fromTexture("characters/spider_baker/hanging", entity)
    sprite:setPosition(entity.x, entity.y)
    sprite:setJustification(0.5, 1.0)
    
    -- Draw web line up
    local webY = entity.webY or -80
    -- Web visual would be handled by game
    
    return sprite
end

function spiderBaker.selection(room, entity)
    return utils.rectangle(entity.x - 10, entity.y - 20, 20, 20)
end

return spiderBaker
